Loki Query Autocomplete: better suggestions and insertions for Logfmt

matyax opened the pull request:
After the recent changes in Logfmt, there were scenarios where we were making suggestions for things that were already in the query, and there were also situations were the user had written a label or was in the process of writing one, and we were interfiring with it without knowing if we can insert a trailing comma or not.

For example:

- `{label="value"} | logfmt `, deleting the last space would trigger a `logfmt` suggestion that would also come with an extra pipe.
- `{label="value"} | logfmt someString` we couldn't tell if `someString` is the label the user wants, or was in the process of writing it.

For the tricky scenarios, we now wait for a trailing space or a trailing comma to make suggestions that we know are the right ones.

Other changes:
- Simplified `getSituation()`.
- Removed `getErrorNode()` in favor of `resolveCursor()`.
- Updated old comments.
- No longer suggest parsers when you're in IN_LOGFMT.
- Supports autocompleting logfmt in the middle of the expression.

**Which issue(s) does this PR fix?**:

Fixes https://github.com/grafana/grafana/issues/75493
Related with https://github.com/grafana/grafana/issues/71121

**Special notes for your reviewer:**

- The autocomplete experience should not have odd behaviors where extra symbols are being inserted. 
- It should also wait for spaces, commas, or pipes to have a better understanding of user intentions.
- Building log queries inside metric queries should be as good as outside of metric queries.

Demo:

https://github.com/grafana/grafana/assets/1069378/3e432099-f268-4629-a230-2d1eaf9f3f56

matyax committed:
Logfmt autocompletion: better handling of trailing commas

gtk-grafana commented on public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.ts:
```diff
@@ -380,14 +380,26 @@ export async function getLogfmtCompletions(
-  const labelPrefix = otherLabels.length === 0 || trailingComma ? '' : ', ';
   const labels = extractedLabelKeys.filter((label) => !otherLabels.includes(label));
+
+  // No other labels or trailing comma, so we don't need to add a prefix
```
It looks like this only works for the first label in the list, on this branch I'm seeing the same behavior on 2nd+ label

https://github.com/grafana/grafana/assets/109082771/912eba47-70af-4eb1-b102-38f6b9c520c5

Start with a query with multiple labels passed into logfmt
`{app_kubernetes_io_name="apisix"} | logfmt  job , name , namespace`
And then reproduce the issue by deleting a few chars from the last label, and then selecting the autocomplete.
`{app_kubernetes_io_name="apisix"} | logfmt job , name , , namespace`

matyax replied:
It's very odd because I tested the same and I can only reproduce the extra pipe that you show in the video, but that is unrelated (which now that we're at it I will try to improve that one as well).

As for the labels:

https://github.com/grafana/grafana/assets/1069378/3ae64a74-a750-40ec-bbff-4d62b8030b0c

gtk-grafana replied:
I checked out the branch again and I'm still seeing this:

https://github.com/grafana/grafana/assets/109082771/dd45694c-1a29-435d-bc41-cb2467354f90

The simplest reproduction is to start with a query with multiple labels, and then delete a few chars from the last label, and then select the autocompletion result.

matyax committed:
Situation: refactor getSituation and  getErrorNode

Looking for simplicity, removing redundancy, and looking for a more accurate node resolution

matyax committed:
Autocompletion: refactor logfmt and top level

matyax committed:
Update tests

matyax commented:
@gtk-grafana approached this from a different perspective, writing queries and paying close attentions not only to the possible situation where the user is, but also to what would be more useful to offer and how we can prevent misinterpretations of user input.

Take this for a spin and let me know how it feels.

gtk-grafana commented:
:( now it doesn't autocomplete logfmt labels if you press space without comma (url in top left, first local, then ops)
https://github.com/grafana/grafana/assets/109082771/b1b9cb3b-bc95-45ab-a424-4d33ffdf2691

Also it seems that autocomplete doesn't work when within a label anymore
https://github.com/grafana/grafana/assets/109082771/ddb7f59a-5ccf-45f7-8359-add4f8819759

matyax commented:
> :( now it doesn't autocomplete logfmt labels if you press space without comma (url in top left, first local, then ops)

It does, just after `logftmt ` or after a comma. It's in the description but let me know if it's not clera.

> Also it seems that autocomplete doesn't work when within a label anymore

Same.

gtk-grafana commented:
No latency, the labels just never show up after a space. I can demo after the standup tomorrow

gtk-grafana commented on public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.test.ts:
```diff
@@ -814,6 +724,21 @@ describe('IN_LOGFMT completions', () => {
     `);
   });
 
+  it('autocompleting logfmt labels should correctly add trailing commas', async () => {
```
```
it('autocompleting logfmt with flags should only offer labels when the user has a trailing space', async () => {
    const situation: Situation = {
      type: 'IN_LOGFMT',
      logQuery: `{job="grafana"} | logfmt label3 `,
      flags: true,
      otherLabels: ['label1'],
    };

    expect(await getCompletions(situation, completionProvider)).toMatchInlineSnapshot(`
      [
        {
          "insertText": "label2",
          "label": "label2",
          "triggerOnInsert": false,
          "type": "LABEL_NAME",
        },
      ]
    `);
  });
```
Fails on this branch if you just delete the comma at the end. Can we add unit tests for all of the test cases we've discovered so far?

matyax replied:
I forgot to update the test titles, so just synced them and added a few more. If there's anything else you think we have to test just let me know.

matyax committed:
Autocompletion: better handling of trailing characters

matyax commented:
Thanks again for the feedback @gtk-grafana . I made several improvements, and we should be close if not finally done.

matyax committed:
Completions: Update data to type changes

matyax commented on public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.test.ts:
```diff
@@ -732,93 +676,73 @@ describe('IN_LOGFMT completions', () => {
+  it('autocompleting logfmt with labels and trailing comma should only return labels', async () => {
     const situation: Situation = {
       type: 'IN_LOGFMT',
       logQuery: `{job="grafana"} | logfmt`,
```
True. It's just a bad example, since that logQuery is used for other purposes in `completions.ts`. Good catch though.

matyax commented on public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.test.ts:
```diff
@@ -732,93 +676,73 @@ describe('IN_LOGFMT completions', () => {
+  it('autocompleting logfmt with labels and trailing comma should only return labels', async () => {
     const situation: Situation = {
       type: 'IN_LOGFMT',
       logQuery: `{job="grafana"} | logfmt`,
```
```suggestion
      logQuery: `{job="grafana"} | logfmt,`,
```

matyax committed:
Update log query in test situation

gtk-grafana commented:
Still seeing some discrepancy between behavior on main on this branch.

1. Seeing fewer autocomplete suggestions for log labels on local when compared to main, for the same datasource
This branch:
<img width="1339" alt="image" src="https://github.com/grafana/grafana/assets/109082771/ab23707b-ecaf-4347-b880-ed2d13953d50">
Main:
<img width="1721" alt="image" src="https://github.com/grafana/grafana/assets/109082771/16411ced-2144-4af4-bbd9-7841191c56e2">

2. On main, autocomplete is working with partial labels, but not on this branch
Main: 
<img width="1728" alt="image" src="https://github.com/grafana/grafana/assets/109082771/f25e97a4-e46c-401d-9fe6-3d374592268f">

This branch:
<img width="1247" alt="image" src="https://github.com/grafana/grafana/assets/109082771/76821f74-e6ae-4da2-a8dc-c24650ff94c6">

Looks like it's inserting the commas correctly through when adding a label in the middle of a list though

matyax committed:
Format code

matyax committed:
Merge branch 'matyax/logfmt-labels-trailing' of github.com:grafana/grafana into matyax/logfmt-labels-trailing

gtk-grafana commented:
Looks like issue #1 above was bad testing on my part. Unable to reproduce that now

matyax committed:
Autocomplete: support offering partial labels

matyax commented:
I was in the process of dumping further improvements into a follow-up issue when I had an epiphany. Let's try again? No rush 🙇

matyax committed:
Completion: expand test cases

gtk-grafana commented on public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.ts:
```diff
@@ -351,43 +351,63 @@ export async function getAfterSelectorCompletions(
 export async function getLogfmtCompletions(
   logQuery: string,
```
if all of these args are properties on the situation object (besides dataProvider), do we want to keep arg-stuffing this function or should we just pass in the situation to this function?

matyax replied:
At least in this PR, yes. I would love to receive situations at some point, but because of how situations have been [defined](https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.ts#L100-L146), the compiler has a hard time deciding which situation it is.

![imagen](https://github.com/grafana/grafana/assets/1069378/b100d130-1ea0-4cca-b4ba-71985c4710bb)

I've recently discussed with the squad that, whenever we have time, I would really like to go back to this implementation and do a round of clean up of the original implementation.

gtk-grafana commented:
Wonderful! Yeah, this works great! All of my test cases are passing when running locally, it is always providing autocomplete results, and doesn't produce invalid queries.

The code needs a bit of clean up still, but functionally this is a huge improvement.

matyax commented:
Thanks @gtk-grafana 

> The code needs a bit of clean up still, but functionally this is a huge improvement.

What do you mean in particular?

gtk-grafana commented on public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.ts:
```diff
@@ -351,43 +351,63 @@ export async function getAfterSelectorCompletions(
-  const labelPrefix = otherLabels.length === 0 || trailingComma ? '' : ', ';
   const labels = extractedLabelKeys.filter((label) => !otherLabels.includes(label));
+
+  /**
+   * {label="value"} | logfmt ^
+   * - trailingSpace: true, trailingComma: false, otherLabels: []
+   * {label="value"} | logfmt lab^
+   * trailingSpace: false, trailignComma: false, otherLabels: [lab]
+   * {label="value"} | logfmt label,^
+   * trailingSpace: false, trailingComma: true, otherLabels: [label]
+   * {label="value"} | logfmt label, ^
+   * trailingSpace: true, trailingComma: true, otherLabels: [label]
+   */
```
Can we delete these comments or add some context?

matyax replied:
To be done in https://github.com/grafana/grafana/pull/76144 along with the rest of the suggestions.

gtk-grafana commented on public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/situation.ts:
```diff
@@ -461,24 +467,38 @@ function resolveLogfmtParser(_: SyntaxNode, text: string, cursorPosition: number
-  // `{x="y"}` situation, with the cursor at the end
-
+  /**
+   * Top level examples:
```
Can you expand on this a bit?

gtk-grafana commented on public/app/plugins/datasource/loki/components/monaco-query-field/monaco-completion-provider/completions.ts:
```diff
@@ -351,43 +351,63 @@ export async function getAfterSelectorCompletions(
+  if (!trailingComma && !flags) {
+    completions = [...LOGFMT_ARGUMENT_COMPLETIONS];
+  }
+  // {label="value"} | logfmt --flag ^
+  // {label="value"} | logfmt label, label2 ^
```
More confusing comments

gtk-grafana reviewed (approved):
Looks good, had a few nits/questions, but this is a huge improvement. Thanks again for all your work on this!
